src/riak_client.erl

```diff
@@ -398,20 +399,45 @@ delete(Bucket,Key,Options,Timeout,{?MODULE, [Node, _ClientId]}=THIS) when is_lis
         true ->
             consistent_delete(Bucket, Key, Options, Timeout, THIS);
         false ->
-            normal_delete(Bucket, Key, Options, Timeout, THIS);
+            maybe_normal_delete(Bucket, Key, Options, Timeout, THIS);
         {error,_}=Err ->
             Err
     end;
 delete(Bucket,Key,RW,Timeout,{?MODULE, [_Node, _ClientId]}=THIS) ->
     delete(Bucket,Key,[{rw, RW}], Timeout, THIS).
 
+%% If this is a write-once bucket, delete without read-before-delete
+%% (write_once_delete).  Else do a normal delete.
+
+maybe_normal_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]}) ->
+    case write_once(Node, Bucket) of
+        true ->
+	    write_once_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]});
+        false ->
+	    normal_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]});
+        {error,_}=Err ->
+            Err
+    end.
+
+%% This version is exported, since it is called directly from
+%% riak_kv_ts_api:delete_data/4
+
+write_once_delete(Bucket, Key, Options, {?MODULE, [Node, ClientId]}) ->
+    write_once_delete(Bucket, Key, Options, ?DEFAULT_TIMEOUT, {?MODULE, [Node, ClientId]}).
+
+%% Write-once get path ignores vclocks.  We pass in an empty vclock
+%% (vclock:fresh()) so that no read-before-delete is triggered in
+%% riak_kv_delete:delete/8
+
+write_once_delete(Bucket, Key, Options, Timeout, {?MODULE, [Node, ClientId]}) ->
+    normal_delete_vclock(Bucket, Key, vclock:fresh(), Options, Timeout, {?MODULE, [Node, ClientId]}).
```

Comment:
Don't forget that vclocks have fairly low timestamp resolution (1 second), so the window of opportunity for collisions is quite large. I don't think that really impacts the decision-making process, but worth noting.

I agree, exporting and using `w1c_vclock/1` seems logical.